Adopt SPDX

jforissier opened the pull request:
- Patch 1 adds SPDX license identifiers to all the source files that have some known licensing terms
- Patches 2 and 3 remove redundant/useless text from files owned by Linaro
- Patch 4 documents the format of the copyright and license headers

jenswi-linaro commented on core/arch/arm/include/arm32.h:
```diff
@@ -1,6 +1,7 @@
 /*
  * Copyright (c) 2016, Linaro Limited
  * Copyright (c) 2014, STMicroelectronics International N.V.
+ * SPDX-License-Identifier: BSD-2-Clause
```
Regardless of one may think of "All rights reserved", but if it's there it should follow the copyright statement directly. License stuff should be below.

jforissier replied:
Agreed. I'll fix my script.

jenswi-linaro commented on core/arch/arm/include/tee/entry_fast.h:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2015, Linaro Limited
+ * SPDX-License-Identifier: BSD-2-Clause
```
Prefer to not break up the copyright statement with this.

jforissier replied:
Will fix.

jenswi-linaro commented on core/arch/arm/kernel/kern.ld.S:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2014, Linaro Limited
+ * SPDX-License-Identifier: BSD-2-Clause
```
The license that follows in this comment could be removed as it only covers the Linaro changes.

The comment with a license below should be left intact though (just as it is now).

jforissier replied:
I agree, but that is a bit more complex to automate. I have to make sure I nuke the license that goes with the Linaro copyright, not the other one(s). That's why I chose to clean only the files that are 100% Linaro copyright.
Maybe I should just take care of this manually.

etienne-lms commented on core/arch/arm/include/arm.h:
```diff
@@ -2,28 +2,6 @@
  * Copyright (c) 2015, Linaro Limited
  * SPDX-License-Identifier: BSD-2-Clause
  * All rights reserved.
- *
```
Does this deserves an update of the date 2015-2017?

jenswi-linaro replied:
I suppose, but I'd rather not review that in this PR as it's quite massive as it is already.

etienne-lms replied:
true :)

etienne-lms commented on documentation/copyright_and_license_headers.md:
```diff
@@ -0,0 +1,65 @@
+      /*
+       * Copyright (c) <year>, <...>
+       *
+       * <License text, BSD 2-Clause>
```
s/2/3/

jforissier replied:
Yup

etienne-lms commented on documentation/copyright_and_license_headers.md:
```diff
@@ -0,0 +1,65 @@
+------------------------------
+
+- (Rule 2.1) SPDX license identifiers shall be added according to the license
+  notice(s) in the file.
```
Is there always an spdx id for an existing imported content? What if not?

jforissier replied:
Not always. For example: https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/kernel/kern.ld.S#L28-L49. I could not find this exact license on the SPDX license page (https://spdx.org/licenses/).
In this case we can't do anything.

etienne-lms replied:
So maybe replace `shall` by `should` in this doc.

jforissier commented:
Force-pushed. Changed:
- Fixed placement of SPDX IDs
- Fixed incorrect detection of BSD-2-Clause in some ARM headers when it should have been BSD-3-Clause
- Fixed doc typo

jforissier commented:
Updated.
- Fixed: use the same style as the current comment block when inserting an SPDX ID (some C files do not repeat the asterisk on each line)
